### ISSUE TITLE
gRPC client doesn't see server failure if request Multi stream is long

### ISSUE THREAD

**Issue by AlekseyMay** [quarkusio/quarkus#31442]
### Describe the bug

Two gRPC apps, client and server. Both using Mutiny Grpc Generation. 
Client is streaming data for several seconds to the server. For example 10000 elements in 3 seconds. Server throws exception at the beginning of the stream, for example at the 30 element. Client doesn't see this failure, it is just stucked with nothing.
If the stream is short and fast all is ok, problem occurs only if the stream is long and takes some time.

I am not a gRPC expert, maybe it is expected behavior, but looks like a bug.

### Expected behavior

Despite the length of the stream and the time when error happens on the server side to see error on the client side.

### Actual behavior

Error is propagated to the client side only if the stream is short and fast.

### How to Reproduce?

Open two projects from archive.
[grpcRepr.zip](https://github.com/quarkusio/quarkus/files/10839009/grpcRepr.zip)

1. start server app
2. start client app

Client app is doing the request on start, so just follow the logs.
To make error propagate again just replace `Multi.createFrom().range(1, 10000)` with `Multi.createFrom().range(1, 50)`

### Output of `java -version`

17

### Quarkus version or git rev

2.16.3

**Comment by cescoffier** [comment]
I believe it's correct. The failure happens once the already emitted item has been sent and processed. However, I will have a look.

**Comment by alesj** [comment]
@AlekseyMay can you move the zip contents into some GitHub repo, and I'll have a look there ...

**Comment by alesj** [comment]
I moved the reproducer here:
* https://github.com/alesj/quarkus/tree/i31442

How is this "test" supposed to fail?
@cescoffier ^

**Comment by alesj** [comment]
@cescoffier btw, with this test I just saw this as well
(on a side not :-))

```
2023-03-09 10:39:24,981 WARNING [io.grp.net.NettyServerHandler] (vert.x-eventloop-thread-0) Exception in onDataRead(): java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:459)
	at io.vertx.grpc.VertxServer$ActualServer.lambda$new$1(VertxServer.java:96)
```

**Comment by alesj** [comment]
@AlekseyMay can you adjust this test so it shows (asserts) your problem ...
* https://github.com/alesj/quarkus/blob/i31442/integration-tests/grpc-streaming/src/test/java/io/quarkus/grpc/example/streaming/LongStreamTestBase.java#L24

**Comment by cescoffier** [comment]
@alesj the test is synchronous. So the multi will emits the numbers and then emit the failure.

**Comment by alesj** [comment]
The test(s) actually fail as expected.
@AlekseyMay @cescoffier ^

When there is a new Vert.x gRPC involved, it does miss actual cause info,
but imho that's the result of the bug @cescoffier just fixed yday -- missing metadata / trailers.

**Comment by cescoffier** [comment]
Yes, the missing data will come as soon as vertx 4.4.1 gets in.

**Comment by alesj** [comment]
What if you modify my tests somehow to reproduce the issue?

**Comment by alesj** [comment]
Hmmm, uh, how is that different — same app vs separate client / server? 🤔

**Comment by cescoffier** [comment]
It does not change anything, there is no "in-process" optimization, and the latency would be the same as if it were in 2 processes running on the same host (things would be different if you used two different machines because of the latency).

**Comment by cescoffier** [comment]
@AlekseyMay trying right now.

**Comment by cescoffier** [comment]
@AlekseyMay @alesj I've found the issue because there is an issue. It's a fun one. 

In a many-to-one operation (as we have here, i.e., the request contains a Multi, and the response is a Uni), if the server sends a failure, we don't cancel the subscription to the request Multi. We need to cancel it, to eventually stop receiving items. "Eventually" is essential here, as the protocol does not guarantee that we won't receive more items, but it _should_ stop emitting. 


BTW, we have the same issue in the many-to-many case.

**Comment by alesj** [comment]
How does this manifest in a separate processes, instead of a single
app+tests?

**Comment by cescoffier** [comment]
The tests were not reproducing the issue, see https://github.com/quarkusio/quarkus/pull/31801. I've modified the tests.

**Comment by cescoffier** [comment]
Fixed by #31801
